### PR TITLE
fix: Return X-WOPI-Lock when a manual lock from outside exists

### LINKED ISSUE #428: Office online does not respect a local locked file by user.
**Describe the bug**
Open a document for local edit.  The document gets locked in the web interface.  Another user can see that it is locked. 

BUT if the user click on it office online opens the document and the user can write in it and changes is save. Not respecting the lock at all. 

If this is done in nextcloud office a popup says that the document is locked and opened in read only mode.  

**To Reproduce**
1. Open file for local edit
2. See that it gets locked
3. Other user that has write permission opens the file
4. writes text and the file gets saved regardless if it is locked for local edit or not.

**Expected behavior**
office online should say that the docment is locked and opened in read only mode

**Screenshots**

**Client details:**
 - OS: linux and windows
 - Browser: chrome and firefox
 - Version: latest stable
 - Device: Desktop and laptop

## Server details

**Operating system**: ubuntu 22.04

**Web server:** apache

**Database:** mysql

**PHP version:** 8.1

**Nextcloud version:** 26.0.0

**Version of the officeonline app** 2.0.1

**Version of Office Online server** 16.0.10396.20000

<details>
<summary>Logs</summary>

#### Nextcloud log (data/nextcloud.log)
```
Nothing relevant found
```

#### Browser log
```
Nothing relevant found 

c) ...
```

</details>

### PR DESCRIPTION
**Pull request by juliushaertl**
Signed-off-by: Julius Härtl <jus@bitgrid.net>

* Resolves #428 
* Target version: main

### Summary

This change makes sure that we return a proper locked response in cases where an outside lock exists. 

I've tested this change with word, excel, powerpoint, however on my OOS server switching to read only mode only work with excel. I couldn't figure out a reason for that yet, but as it seems to work fine with excel it would rather assume this is a bug or misconfiguration of my OOS server.

### Checklist

- [ ] Code is properly formatted
- [ ] Sign-off message is added to all commits
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by juliushaertl** [comment]
Thanks a lot already. Just to double check does switching to the read only mode work for you for all file types (word/excel/PowerPoint)?

**Comment by enoch85** [comment]
Powerpoint:

Locking works, but not read only

image
image

Excel:
Does not even open as it seems (waited ~3 minutes):

image

More testing:
Now I only get internal server error...

```
{"reqId":"iDvMUGWEPhm3iVV7KzGl","level":3,"time":"2023-05-11T10:50:50+00:00","remoteAddr":"1.2.3.4","user":"Jorgen@email.com","app":"officeonline","method":"GET","url":"/apps/officeonline/public?shareToken=&fileName=%252FKollektiv%252FIT%2520S%25C3%25A4kerhetssamordning%252FAP-lista.xlsx&requesttoken=WWRb6zCt%2FyiomUicCRznI3bETeqWA0zp%2B7SSaAdlgQ8%3D%3AKgk0s2bFsF7jyTzPTl2yQiCrK676eTaFyeX7GW1ds2Y%3D&fileId=%7BfileId%7D","message":"Den beg\u00e4rda delningen finns inte mer","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 Edg/112.0.1722.64","version":"26.0.1.1","exception":{"Exception":"OCP\\Share\\Exceptions\\ShareNotFound","Message":"Den beg\u00e4rda delningen finns inte mer","Code":0,"Trace":[{"file":"/srv/www/nextcloud/apps/officeonline/lib/Controller/DocumentController.php","line":349,"function":"getShareByToken","class":"OC\\Share20\\Manager","type":"->"},{"file":"/srv/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"publicPage","class":"OCA\\Officeonline\\Controller\\DocumentController","type":"->"},{"file":"/srv/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/srv/www/nextcloud/lib/private/AppFramework/App.php","line":183,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/srv/www/nextcloud/lib/private/Route/Router.php","line":315,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/srv/www/nextcloud/lib/base.php","line":1056,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/srv/www/nextcloud/index.php","line":36,"function":"handleRequest","class":"OC","type":"::"}],"File":"/srv/www/nextcloud/lib/private/Share20/Manager.php","Line":1537,"Hint":"Den beg\u00e4rda delningen finns inte mer","CustomMessage":"--"}}
{"reqId":"Ov03BXGfHe67oJemiEsr","level":3,"time":"2023-05-11T11:43:48+00:00","remoteAddr":"1.2.3.4","user":"daniel.hansson@email.com","app":"webdav","method":"POST","url":"/remote.php/dav/bulk","message":"Unknown error while seeking content","userAgent":"Mozilla/5.0 (Windows) mirall/3.8.1stable-Win64 (build 20230419) (Nextcloud, windows-10.0.19045 ClientArchitecture: x86_64 OsArchitecture: x86_64)","version":"26.0.1.1","exception":{"Exception":"Sabre\\DAV\\Exception","Message":"Unknown error while seeking content","Code":500,"Trace":[{"file":"/srv/www/nextcloud/apps/dav/lib/BulkUpload/MultipartRequestParser.php","line":129,"function":"isAt","class":"OCA\\DAV\\BulkUpload\\MultipartRequestParser","type":"->"},{"file":"/srv/www/nextcloud/apps/dav/lib/BulkUpload/BulkUploadPlugin.php","line":71,"function":"isAtLastBoundary","class":"OCA\\DAV\\BulkUpload\\MultipartRequestParser","type":"->"},{"file":"/srv/www/nextcloud/3rdparty/sabre/event/lib/WildcardEmitterTrait.php","line":89,"function":"httpPost","class":"OCA\\DAV\\BulkUpload\\BulkUploadPlugin","type":"->"},{"file":"/srv/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":472,"function":"emit","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/srv/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":253,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/srv/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php","line":321,"function":"start","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/srv/www/nextcloud/apps/dav/lib/Server.php","line":366,"function":"exec","class":"Sabre\\DAV\\Server","type":"->"},{"file":"/srv/www/nextcloud/apps/dav/appinfo/v2/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->"},{"file":"/srv/www/nextcloud/remote.php","line":172,"args":["/srv/www/nextcloud/apps/dav/appinfo/v2/remote.php"],"function":"require_once"}],"File":"/srv/www/nextcloud/apps/dav/lib/BulkUpload/MultipartRequestParser.php","Line":111,"message":"Unknown error while seeking content","exception":{},"CustomMessage":"Unknown error while seeking content"}}
{"reqId":"wwtKXR1rrbE9X9aDue8H","level":4,"time":"2023-05-11T12:22:24+00:00","remoteAddr":"1.2.3.4","user":"--","app":"user_saml","method":"POST","url":"/apps/user_saml/saml/acs","message":"invalid_response","userAgent":"Mozilla/5.0 (iOS) NextcloudNotes/4.1.7","version":"26.0.1.1","data":{"app":"user_saml"}}
{"reqId":"wwtKXR1rrbE9X9aDue8H","level":4,"time":"2023-05-11T12:22:24+00:00","remoteAddr":"1.2.3.4","user":"--","app":"user_saml","method":"POST","url":"/apps/user_saml/saml/acs","message":"The status code of the Response was not Success, was Responder -&gt; urn:oasis:names:tc:SAML:2.0:status:RequestDenied","userAgent":"Mozilla/5.0 (iOS) NextcloudNotes/4.1.7","version":"26.0.1.1","data":{"app":"user_saml"}}
{"reqId":"fqQV3UA8LdECOzhtjWst","level":3,"time":"2023-05-11T14:07:10+00:00","remoteAddr":"1.2.3.4","user":"danh","app":"no app in context","method":"POST","url":"/ocs/v2.php/apps/files_sharing/api/v1/shares","message":"Path contains files shared with you","userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/112.0","version":"26.0.1.1","exception":{"Exception":"InvalidArgumentException","Message":"Path contains files shared with you","Code":0,"Trace":[{"file":"/srv/www/nextcloud/lib/private/Share20/Manager.php","line":727,"function":"pathCreateChecks","class":"OC\\Share20\\Manager","type":"->"},{"file":"/srv/www/nextcloud/apps/files_sharing/lib/Controller/ShareAPIController.php","line":720,"function":"createShare","class":"OC\\Share20\\Manager","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/srv/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":230,"function":"createShare","class":"OCA\\Files_Sharing\\Controller\\ShareAPIController","type":"->","args":["*** sensitive parameters replaced ***"]},{"file":"/srv/www/nextcloud/lib/private/AppFramework/Http/Dispatcher.php","line":137,"function":"executeController","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/srv/www/nextcloud/lib/private/AppFramework/App.php","line":183,"function":"dispatch","class":"OC\\AppFramework\\Http\\Dispatcher","type":"->"},{"file":"/srv/www/nextcloud/lib/private/Route/Router.php","line":315,"function":"main","class":"OC\\AppFramework\\App","type":"::"},{"file":"/srv/www/nextcloud/ocs/v1.php","line":64,"function":"match","class":"OC\\Route\\Router","type":"->"},{"file":"/srv/www/nextcloud/ocs/v2.php","line":23,"args":["/srv/www/nextcloud/ocs/v1.php"],"function":"require_once"}],"File":"/srv/www/nextcloud/lib/private/Share20/Manager.php","Line":690,"CustomMessage":"--"}}




```

**Comment by enoch85** [comment]
Also noted, the actual internal server error doesn't get logged anywhere afaik.

**Comment by juliushaertl** [comment]
The log messages indeed don't look related.

> Now I only get internal server error...

How do you trigger that and where you see it? Is there a request id mentioned in the output that could help find the log messages?

**Comment by juliushaertl** [comment]
Regarding the error that powerpoint shows, are you able to find something meaninful regaring the session id in the OOS server logs? I had a hard time checking those and also haven't noticed any message on mine that would indicate an error.

**Comment by juliushaertl** [comment]
Had one more confirmation that this fixes the issue as expected.

**Comment by enoch85** [comment]
> How do you trigger that and where you see it? Is there a request id mentioned in the output that could help find the log messages?

I just try to open the document in the web as usual. Might have been something temporary.

> Regarding the error that powerpoint shows, are you able to find something meaninful regaring the session id in the OOS server logs?

Does OSS have seperate sevrer logs? :man_shrugging:

**Comment by juliushaertl** [comment]
Yes, see https://learn.microsoft.com/en-us/compliance/regulatory/gdpr-for-office-online-server and https://sharepoint.stackexchange.com/questions/286061/where-is-oos-loglocation-path-for-win2019

But I haven't found any useful message in there on my setup

File: lib/Controller/WopiController.php [hunk 1]
```diff
@@ -671,6 +671,16 @@ public function postFile($fileId, $access_token) {
 		$wopiOverride = $this->request->getHeader('X-WOPI-Override');
 
 		if ($this->lockManager->isLockProviderAvailable()) {
+			$locks = $this->lockManager->getLocks($wopi->getFileid());
+			$existingLock = array_pop($locks);
```
Comment (blizzz) [inline lib/Controller/WopiController.php:675]: other locks would not be relevant by principle?

Comment (juliushaertl) [reply lib/Controller/WopiController.php:675]: There is currently just the possibility of one lock the way how files_lock app handles them. In the future multiple locks might be possible, if we support locking folders as well. I'll switch to array_shift instead as we are only interested in the closest lock to the file.

Comment (juliushaertl) [reply lib/Controller/WopiController.php]: Adjusted

**Review by blizzz — APPROVED** [review]

**Pull request merged**